File: Source/FcitxCjkInput/InputRecovery.cs
using System;

namespace FcitxCjkInput {
    internal sealed class CommittedCharacterTracker {
        private ControlToken _target;
        private string _characters = "";
        private int _index;
        private int _length;
        private int _expiresAfterFrame = -1;

        public void Expect(ControlToken target, string text, int insertedLength, int frame) {
            var length = Math.Max(0, Math.Min(insertedLength, text.Length));
            if (target.Id == 0 || length == 0)
                return;

            var expected = text.Substring(0, length);
            if (!_target.Equals(target) || frame > _expiresAfterFrame || _index >= _length) {
                _target = target;
                _characters = expected;
                _index = 0;
                _length = expected.Length;
            } else {
                _characters = _characters.Substring(_index, _length - _index) + expected;
                _index = 0;
                _length = _characters.Length;
            }
            _expiresAfterFrame = frame + 1;
        }

        public bool ShouldSuppress(ControlToken target, char character, int frame) {
            if (character == '\0')
                return false;
            if (!_target.Equals(target) || frame > _expiresAfterFrame || _index >= _length) {
                Clear();
                return false;
            }
            if (_characters[_index] != character) {
                Clear();
                return false;
            }

            _index++;
            if (_index >= _length)
                Clear();
            return true;
        }

        public void Clear() {
            _target = default;
            _characters = "";
            _index = 0;
            _length = 0;
            _expiresAfterFrame = -1;
        }
    }

    internal sealed class DirectionalKeyState {
        private const int A = 1 << 0;
        private const int D = 1 << 1;
        private const int S = 1 << 2;
        private const int W = 1 << 3;

        private int _pressed;

        public void Update(int keyValue, bool release) {
            var mask = MaskFor(keyValue);
            if (release)
                _pressed &= ~mask;
            else
                _pressed |= mask;
        }

        public bool IsDown(int keyCode) {
            var mask = MaskFor(keyCode);
            return mask != 0 && (_pressed & mask) != 0;
        }

        public void Clear() {
            _pressed = 0;
        }

        private static int MaskFor(int keyValue) {
            switch (keyValue) {
                case 'a':
                case 'A':
                    return A;
                case 'd':
                case 'D':
                    return D;
                case 's':
                case 'S':
                    return S;
                case 'w':
                case 'W':
                    return W;
                default:
                    return 0;
            }
        }
    }

    internal static class GameplayKeyRecovery {
        public static bool ShouldRecover(bool original, bool textFieldActive, bool cameraDolly,
            int primaryKey, int secondaryKey, DirectionalKeyState keys) {
            return original || (!textFieldActive && cameraDolly &&
                (keys.IsDown(primaryKey) || keys.IsDown(secondaryKey)));
        }
    }
}

File: Source/FcitxCjkInput/InputRecovery.fs
namespace FcitxCjkInput

open System

type CommittedCharacterTracker() =
    let mutable target = Unchecked.defaultof<ControlToken>
    let mutable characters = ""
    let mutable index = 0
    let mutable length = 0
    let mutable expiresAfterFrame = -1

    member this.Expect(newTarget: ControlToken, text: string, insertedLength: int, frame: int) =
        let len = max 0 (min insertedLength text.Length)

        if newTarget.Id = 0 || len = 0 then
            ()
        else

            let expected = text.Substring(0, len)

            if not (target.Equals(newTarget)) || frame > expiresAfterFrame || index >= length then
                target <- newTarget
                characters <- expected
                index <- 0
                length <- expected.Length
            else
                characters <- characters.Substring(index, length - index) + expected
                index <- 0
                length <- characters.Length

            expiresAfterFrame <- frame + 1

    member this.ShouldSuppress(checkTarget: ControlToken, character: char, frame: int) =
        if character = '\000' then
            false
        elif not (target.Equals(checkTarget)) || frame > expiresAfterFrame || index >= length then
            target <- Unchecked.defaultof<ControlToken>
            characters <- ""
            index <- 0
            length <- 0
            expiresAfterFrame <- -1
            false
        elif characters.[index] <> character then
            target <- Unchecked.defaultof<ControlToken>
            characters <- ""
            index <- 0
            length <- 0
            expiresAfterFrame <- -1
            false
        else
            index <- index + 1

            if index >= length then
                target <- Unchecked.defaultof<ControlToken>
                characters <- ""
                index <- 0
                length <- 0
                expiresAfterFrame <- -1

            true

    member _.Clear() =
        target <- Unchecked.defaultof<ControlToken>
        characters <- ""
        index <- 0
        length <- 0
        expiresAfterFrame <- -1

type DirectionalKeyState() =
    [<Literal>]
    static let A = 1 <<< 0

    [<Literal>]
    static let D = 1 <<< 1

    [<Literal>]
    static let S = 1 <<< 2

    [<Literal>]
    static let W = 1 <<< 3

    let mutable pressed = 0

    member _.Update(keyValue: int, release: bool) =
        let mask = DirectionalKeyState.MaskFor keyValue

        if release then
            pressed <- pressed &&& ~~~mask
        else
            pressed <- pressed ||| mask

    member _.IsDown(keyCode: int) =
        let mask = DirectionalKeyState.MaskFor keyCode
        mask <> 0 && (pressed &&& mask) <> 0

    member _.Clear() = pressed <- 0

    static member private MaskFor(keyValue: int) =
        match keyValue with
        | 97
        | 65 -> A
        | 100
        | 68 -> D
        | 115
        | 83 -> S
        | 119
        | 87 -> W
        | _ -> 0

module GameplayKeyRecovery =
    let shouldRecover
        (original: bool)
        (textFieldActive: bool)
        (cameraDolly: bool)
        (primaryKey: int)
        (secondaryKey: int)
        (keys: DirectionalKeyState)
        =
        original
        || (not textFieldActive
            && cameraDolly
            && (keys.IsDown(primaryKey) || keys.IsDown(secondaryKey)))
